### PR TITLE
Fix keras 3 custom object serialization

### PR DESCRIPTION
**Pull request by serena-ruan**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/serena-ruan/mlflow/pull/11257?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11257/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11257
```

</p>
</details>

### Related Issues/PRs

Resolve  [Failing tensorflow models test](https://github.com/mlflow-automation/mlflow/actions/runs/8049486027/job/21994075264)

### How is this PR tested?

- [x] Existing unit/integration tests
- [ ] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [x] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by serena-ruan** [comment]
Weird: tested in devbox on nightly build the tests pass: 
```
========================================================================== slowest 10 durations ===========================================================================
13.74s call     tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_single_multidim_input_model_spark_udf[local]
3.07s setup    tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_single_multidim_input_model_spark_udf[local]
0.93s teardown tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_single_multidim_input_model_spark_udf[local]
===================================================================== 1 passed, 10 warnings in 19.86s ===================================================================

(mlflow) serena.ruan@ip-10-110-26-100:~/repos/mlflow$ pip list | grep keras
keras-nightly                 3.1.0.dev2024022703
(mlflow) serena.ruan@ip-10-110-26-100:~/repos/mlflow$ pip list | grep tf
platformdirs                  4.2.0
pyarrow-hotfix                0.6
setfit                        1.0.3
tf-nightly                    2.17.0.dev20240226
tf2onnx                       1.16.1
```

**Comment by serena-ruan** [comment]
These fail on CI again while they pass in my devbox:
```
========================================================================== slowest 10 durations ===========================================================================
13.59s call     tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_single_multidim_input_model_spark_udf[local]
11.11s call     tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_scoring_server_successfully_on_multi_multidim_input_model
11.08s call     tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_scoring_server_successfully_on_single_multidim_input_model
4.46s call     tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_multi_multidim_input_model_spark_udf[local]
3.22s setup    tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_single_multidim_input_model_spark_udf[local]
0.83s teardown tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_scoring_server_successfully_on_multi_multidim_input_model
0.20s setup    tests/tensorflow/test_keras_pyfunc_model_works_with_all_input_types.py::test_multi_multidim_input_model_spark_udf[local]

(3 durations < 0.005s hidden.  Use -vv to show these durations.)
===================================================================== 4 passed, 13 warnings in 46.62s ===================================================================
```

**Comment by serena-ruan** [comment]
I followed CI process to create a new conda environment that's exactly the same as CI, and tests still pass in devbox

**Comment by serena-ruan** [comment]
passed! https://github.com/mlflow/mlflow/actions/runs/8065640228/job/22031908694?pr=11257

**Review by B-Step62 — APPROVED** [review]
LGTM, thanks for the fix! Just for curiosity how this import change relates to the serialization issue...?

File: mlflow/tensorflow/__init__.py [hunk 1]
```diff
@@ -121,9 +121,9 @@ def get_global_custom_objects():
         A live reference to the global dictionary of custom objects.
     """
     try:
-        from tensorflow import keras
+        from tensorflow.keras.saving import get_custom_objects
 
-        return keras.saving.get_custom_objects()
+        return get_custom_objects()
```
[the remark this answers is not in the document]
Comment (serena-ruan) [reply mlflow/tensorflow/__init__.py:126]: It's weird, but the old importing way cause error:
```
>>> from tensorflow import keras
>>> keras.saving.get_custom_objects()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/serena.ruan/miniconda3/envs/mlflow-ci/lib/python3.9/site-packages/tensorflow/python/util/lazy_loader.py", line 211, in __getattr__
    return getattr(module, item)
AttributeError: module 'keras._tf_keras.keras' has no attribute 'saving'
```

**Pull request merged**